app/assets/javascripts/data_sets/editTable.js.coffee

```diff
@@ -17,148 +17,239 @@ uploadSettings =
-
-  view.onRowCountChanged.subscribe (e, args) ->
-    grid.updateRowCount()
-    grid.render()
-
-  view.onRowsChanged.subscribe (e, args) ->
-    grid.invalidateRow args.rows
-    grid.render()
-
-  $('.edit_table_add').click ->
-    if $('.edit_table_save').hasClass 'disabled'
-      return
-
-    grid.getEditorLock().commitCurrentEdit()
-    grid.resetActiveCell()
+Grid = (cols, data, submit) ->
+  # slickgrid objects
+  view = null
+  grid = null
+
+  # objects relevant to popovers
+  validationPop = null
+  validationPopMsg = null
+  deletionPop = null
+  deletionPopTimer = null
+
+  # ID counter for rows
+  currID = 0
+
+  # queue of save and delete row actions to execute on validation
+  actions = []
+
+  initialize = ->
+    # add the delete button to each row
+    cols.push
+      id: 'del'
+      field: 'del'
+      name: ''
+      width: 0
+      formatter: (row, cell, value, columnDef, dataContext) ->
+        "<i class='fa fa-close slick-delete'></i>"
+
+    # slickgrid's grid options
+    options =
+      autoEdit: true
+      editable: true
+      enableCellNavigation: true
+      enableColumnReorder: false
+      forceFitColumns: true
+      rowHeight: 35
+      syncColumnCellResize: true
+
+    # associate IDs with each row in the table
+    for _, i in data
+      data[i]['id'] = currID
+      currID += 1
+
+    # create the required slickgrid objects, give slickgrid the data
+    view = new Slick.Data.DataView()
+    grid = new Slick.Grid '#slickgrid-container', view, cols, options
+    view.setItems data
+
+  subscribeEvents = ->
+    $(window).resize resizeWindow
+
+    $(document).click (e) ->
+      if $(e.target).closest('.slick-row, #dt-picker').length == 0
+        grid.getEditorLock().commitCurrentEdit()
+
+    $('.edit_table_add').click ->
+      if $('.edit_table_save').hasClass 'disabled'
+        return
+      addRow()
+
+    $('.edit_table_cancel').click ->
+      projectID = $(document).data 'project'
+      window.location = "/projects/#{projectID}"
+
+    $('.edit_table_save').click queueSaveGrid
+
+    view.onRowsChanged.subscribe (e, args) ->
+      grid.invalidateRow args.rows
+      grid.render()
+
+    view.onRowCountChanged.subscribe (e, args) ->
+      grid.updateRowCount()
+      grid.render()
+
+    grid.onClick.subscribe (e, args) ->
+      cell = grid.getCellFromEvent e
+      if cell.cell == grid.getColumns().length - 1
+        queueDeleteRow(cell.row)
+        if grid.getDataLength() == 0
+          addRow()
+
+    grid.onAfterCellEditorDestroy.subscribe ->
+      hidePopover()
+      processActions()
+
+    grid.onValidationError.subscribe (e, args) ->
+      showPopover $(args.cellNode), args.validationResults.msg
+      actions = []
+
+    grid.onKeyDown.subscribe (e, args) ->
+      if e.keyCode == 13 and grid.getDataLength() - 1 == grid.getActiveCell().row
+        addRow()
+
+    $(window).trigger 'resize'
+
+    setTimeout ->
+      $('.slick-cell.l0.r0').first().trigger 'click'
+    , 1
```

Comment:
is there a reason this can't be on the line above?

Also, aside from function definitions, the optional parens can help to group arguments neatly